### PR TITLE
[CodeGen] Allow mixed scalar type constraints for inline asm

### PR DESCRIPTION
**Pull request by dfszabo**
GCC supports code like "asm volatile ("" : "=r" (i) : "0" (f))" where i is integer type and f is floating point type. Currently this code produces an error with Clang. The change allows mixed scalar types between input and output constraints.

### CODE REVIEW

**Comment by RKSimon** [comment]
Are there existing bug issues associated with this?

**Comment by dfszabo** [comment]
> Are there existing bug issues associated with this?

I don't know. Found this issue in our downstream target.

File: clang/test/CodeGen/inline-asm-fp-to-int.c [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// RUN: %clang_cc1 %s -emit-llvm -o /dev/null
+
+unsigned test(float f)
```
Comment (topperc) [inline clang/test/CodeGen/inline-asm-fp-to-int.c:3]: What IR does this generate?

Comment (topperc) [reply clang/test/CodeGen/inline-asm-fp-to-int.c:3]: This test is using -emit-llvm, but the patch is in SelectionDAG. SelectionDAG doesn't run with -emit-llvm

Comment (dfszabo) [reply clang/test/CodeGen/inline-asm-fp-to-int.c]: Thanks for the review. You are right, indeed the test was pointless, the problem is with CodeGen. Updated the issue accordingly.

**Comment by dfszabo** [comment]
Ping

**Comment by dfszabo** [comment]
Ping

File: llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+; CHECK-NEXT:    #APP
+; CHECK-NEXT:    #NO_APP
+; CHECK-NEXT:    movl %eax, -{{[0-9]+}}(%rsp)
+; CHECK-NEXT:    retq
+entry:
+  %f.addr = alloca float, align 4
+  %i = alloca i32, align 4
+  store float %f, ptr %f.addr, align 4
+  %0 = load float, ptr %f.addr, align 4
+  %1 = call i32 asm sideeffect "", "=r,0,~{dirflag},~{fpsr},~{flags}"(float %0)
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:28]: Use named values in tests

File: llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  store i32 %1, ptr %i, align 4
+  %2 = load i32, ptr %i, align 4
+  ret i32 %2
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:32]: Can you also try the same, except with a mixed pointer and int/float? Also some vector cases?

**Review by arsenm — CHANGES_REQUESTED** [review]

File: llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+; CHECK-NEXT:    #APP
+; CHECK-NEXT:    #NO_APP
+; CHECK-NEXT:    movl %eax, -{{[0-9]+}}(%rsp)
+; CHECK-NEXT:    # kill: def $eax killed $eax killed $rax
+; CHECK-NEXT:    retq
+entry:
+  %f.addr = alloca float*, align 4
+  %i = alloca i32, align 4
+  store float* %f, ptr %f.addr, align 4
+  %load_f = load float*, ptr %f.addr, align 4
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:48]: Test needs to be updated to use opaque pointers. Also you don't need all of this intermediate alloca stuff, you can simplify the incoming values and uses

**Review by arsenm — CHANGES_REQUESTED** [review]
Test needs cleanup

File: llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+; C source used for generating this test:
+
+; unsigned test(float f)
+; {
+;    unsigned i;
+;    asm volatile ("" : "=r" (i) : "0" (f));
+;    return i;
+; }
+
+
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:13]: Add some comments explaining what this is showing

File: llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+define i32 @test_int_ptr(ptr %f) {
+; CHECK-LABEL: test_int_ptr:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    movq %rdi, %rax
+; CHECK-NEXT:    #APP
+; CHECK-NEXT:    #NO_APP
+; CHECK-NEXT:    # kill: def $eax killed $eax killed $rax
+; CHECK-NEXT:    retq
+entry:
+  %asm_call = call i32 asm sideeffect "", "=r,0,~{dirflag},~{fpsr},~{flags}"(ptr %f)
```
Comment (arsenm) [inline llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:36]: Is it really allowed to have the mismatched sizes, pointer 64 with i32? Should that be an x86-only thing?

Comment (arsenm) [reply llvm/test/CodeGen/X86/inline-asm-int-to-fp.ll:36]: What about double + i32? Or i16 + float?

**Review by arsenm — APPROVED** [review]

**Comment by dfszabo** [comment]
Well if approved, then can someone merge it?

**Pull request merged**